iemgui: merge radio and slider code

umlaeute opened the pull request:
this merges the implementation of sliders and radiobuttons.

the vertical/horizontal variants of these widgets are almost identical (from an implementation POV), but having them in two distinct files each, makes bug-fixing a brittle task (in general you keep forgetting doing the same thing in both files).

so this PR get's rid of a lot of duplicate code in the implementation of hslider/vslider and hradio/vradio, by merging them into `g_slider.c` and `g_radio.c` (no more `g_*dial.` :stuck_out_tongue: )

since the merge changed a lot of lines (mostly renaming stuff, but also adding code for the horizontal resp. vertical cases), this breaks some parts of #1628.
in return i have started to implement the Tcl/Tk `canvas move` semantics for faster displacements of the iemguis (similar to @porres solution, this breaks the ABI of the iemgui objects, but (unlike #1628) it keeps API compatibility. i think there's only a single external out there that uses `g_all_guis.h` (@tmusil's *iemgui*), but i really think that these headers should never have been made public in the first place).

this PR was inspired by https://github.com/pure-data/pure-data/issues/1344#issuecomment-937001815 (@millerpuckette 's request to unify sliders and radiobuttons)


- Closes: #632 
- Closes: #920 
- Closes: #992 
- Closes: #1021 
- Closes: #1022 
- Closes: #1042 
- Closes: #1043 
- Closes: #1121 
- Closes: #1583 
- Closes: #1629 
- Closes: #1637 
- Closes: #1638 
- Closes: #1646

umlaeute commented:
the PR is based on `develop` (hence the number of commits)

porres commented:
> i really think that these headers should never have been made public in the first place).

Agreed!

umlaeute commented:
also there's a new `orientation` message, that allows you to change the horizontal/vertical orientation on-the-fly (but i haven't bothered to create a dialog entry for this).

umlaeute commented:
@porres: the refactoring now uses the same class for both variants, which means that we only have a single helpfile for `slider` resp `radio`. i'm currently using the `hradio-help` resp. `hslider-help`, but you are very welcome to provide a unified help-patch (e.g. `slider-help.pd` resp `radio-help.pd`).

millerpuckette commented:
This is an excellent thing to do.  I was intending to do it (someday) but am happy you got in front of me.  I had also planned to include toggles which are also functionally the same as sliders/radios.

porres commented:
> I had also planned to include toggles which are also functionally the same as sliders/radios.

I remember you wanted to include a new GUI to replace sliders/radios, but it would also be a 'toggle'? :) really curious about this

sebshader commented on src/g_radio.c:
```diff
@@ -48,113 +42,90 @@ void hradio_draw_update(t_gobj *client, t_glist *glist)
-             xx11b, yy11,
-             xx11b + iow, yy11 - IEMGUI_ZOOM(x) + ioh);
+    int dx = text_xpix(&x->x_gui.x_obj, glist) - x->x_gui.x_prevX;
+    int dy = text_ypix(&x->x_gui.x_obj, glist) - x->x_gui.x_prevY;
```
I'm curious and this is somewhat OT, but when should we consider previous tcl/tk versions deprecated? obviously 8.4 is now, but it would be great to use the canvas "moveto" command introduced in 8.6 to avoid any unnecessary arithmetic in the audio thread (even if it is just 2 subtractions right now)..

sebshader replied:
also, should the other iemguis also be updated to use "move" in this pr as well?

umlaeute replied:
if you manage to get Tcl/Tk to work on OSX-10.6/PowerPC, be my guest.

in the meantime, there are *many more* unnecessary calculations in the audio-thread that might be lifted to the GUI.

sebshader replied:
.. my osx powerpc's hd went the way of the dodo ~7 yrs ago right after adobe discontinued flash support for it and then melted in a fire so I guess not

umlaeute replied:
on rethinking, i now believe that `moveto` is actually harmful: it assumes that the anchor point of the widget (which will eventually be positioned on the location declared by `moveto`) is the top-left corner of the bounding box.

that felt like a bad idea (conceptually), but i thought it would only be a theoretical problem, until i thought about *labels*.

umlaeute replied:
> also, should the other iemguis also be updated to use `move` in this pr as well?

yes.

sebshader replied:
I guess that makes sense. hopefully in the future most click/edit handling will happen in the gui process anyways..

umlaeute commented:
i just noticed that this currently changes z-ordering whenever the dialog updates a GUI-object.
i'll submit a fix shortly...(and un-draft the PR)

porres commented:
I closed my PR that did also use "move" for iemguis, opened a new one to use "move" for the other text objects here https://github.com/pure-data/pure-data/pull/1635

But you're welcome to "grab" those changes and include them here, perhaps with some changes of your own, then I'd also close the PR in favor of this. But I would like to highlight that both of them should be merged together, there's no point in merging just one ;)

cheers

umlaeute commented:
i've fixed the z-ordering regression (and a couple of others i discovered while testing..thanks @porres for the new help-patches, it makes testing most features quite easy).

i've also switched the remaining iemgui's to the new `canvas move` code (and did some more housekeeping, like having a consistent ordering of the draw-functions...).

porres commented:
Hi, since you're cleaning up a lot of the code and working mostly in the slider/radio buttons, and since you already cherry picked a fix from a another PR, here are some other fixes that I highly suggest that you also cherry pick and include here ;)

draw vradio's outlet correctly => https://github.com/pure-data/pure-data/commit/a28e85c1dd826dc7a43ccb9492fbd1f22dc8f7a2

other couple of fixes to radio buttons => https://github.com/pure-data/pure-data/pull/831

slider fixes => https://github.com/pure-data/pure-data/pull/933

porres commented:
hmm, here's some more suggestions:

you may also cherry pick this one => https://github.com/pure-data/pure-data/pull/946

This issue is related to the fix you cherry picked from my other PR, see if you can also take care of this then please =>  https://github.com/pure-data/pure-data/issues/1164

and maybe, just maybe, you wizard could find us a solution for this :) => https://github.com/pure-data/pure-data/issues/622

I have been gathering iemguis related stuff in https://github.com/pure-data/pure-data/issues/947 but these are the ones that I think are better suited to be included here on this PR, which I guess should be renamed as it's also cleaning up and doing more things than just merging sliders and radio buttons.

danomatika commented:
This is good. Also the idea of adopting canvas move which is something I've wanted to look into adding for a while now.

porres commented:
Another suggestion to look at in this PR: tcl/tk errors in radio buttons --> https://github.com/pure-data/pure-data/issues/1583

umlaeute commented:
> Another suggestion to look at in this PR: tcl/tk errors in radio buttons --> #1583

as this PR is based on develop, it already includes 885a7cdffe081ccc008330a81409ca39b07a0ff7 which Closes #1583

umlaeute commented:
https://github.com/pure-data/pure-data/commit/a28e85c1dd826dc7a43ccb9492fbd1f22dc8f7a2 a similar fix has already been applied.
@porres would it be possible to check this PR on your local machine to see which issues have already been fixed? (i would also welcome some more testing :-))

porres commented:
> @porres would it be possible to check this PR on your local machine to see which issues have already been fixed? (i would also welcome some more testing :-))

sure thing!

porres commented:
what about this one? --> https://github.com/pure-data/pure-data/issues/1561

porres commented:
built it and testing it, when trying the orientation message and changing from horizontal to vertical I see an alignment issue. 

horizontal:

![Screen Shot 2022-04-21 at 19 15 15](https://user-images.githubusercontent.com/5360142/164560431-7f626cbd-446f-4463-b206-ff4a4b5f11b9.png)

now when changing to vertical

![Screen Shot 2022-04-21 at 19 15 33](https://user-images.githubusercontent.com/5360142/164560506-70496d90-a682-4dcd-a952-de9baa11a189.png)

I can see that the vertical version gets unaligned, and it moves 3 pixels to the right and 2 pixels above

porres commented:
I can confirm https://github.com/pure-data/pure-data/issues/1021 is closed by https://github.com/pure-data/pure-data/commit/a28e85c1dd826dc7a43ccb9492fbd1f22dc8f7a2

and sorry, I missed that https://github.com/pure-data/pure-data/issues/1583 already had a fix

porres commented:
I can confirm this also closes https://github.com/pure-data/pure-data/issues/730

porres commented:
Well, just so we keep track of things,  I also suggested this PR could take care of https://github.com/pure-data/pure-data/pull/933 / https://github.com/pure-data/pure-data/pull/946 / https://github.com/pure-data/pure-data/issues/1164 / and https://github.com/pure-data/pure-data/issues/1561

Will test if they are taken care of already, but wouldn't you know @umlaeute if you had already fixed those in here?

porres commented:
I made patches to be merged in this PR, hope it works.

porres commented:
I tried creating [nbx] with a bigger font, the width is adjusted, but not the height, it seems you also need to set the height according to the init font size.

![Screen Shot 2022-04-21 at 21 02 01](https://user-images.githubusercontent.com/5360142/164569948-01ebe906-6b0b-4fde-9e4a-da5e95e992db.png)

porres commented:
I'd suggest we could do something like alt+click and change the orientation of sliders/radios, what do you think?

porres commented:
> Well, just so we keep track of things, I also suggested this PR could take care of https://github.com/pure-data/pure-data/pull/933 / https://github.com/pure-data/pure-data/pull/946 / https://github.com/pure-data/pure-data/issues/1164 / and https://github.com/pure-data/pure-data/issues/1561

https://github.com/pure-data/pure-data/issues/1164 is not fixed in this PR and has no proposed fix anywhere so far

umlaeute commented:
> I can confirm this also closes https://github.com/pure-data/pure-data/issues/730

no. #730 (and its duplicate #1347) were closed by https://github.com/pure-data/pure-data/commit/e67a91aeb15125e7c811db8fdd3326539c79fd5c which was released in Pd-0.52-1

Ant1r commented:
> slider fixes => #933

yeah, too bad #933 had not been merged before. Now #1042, #1043 and #1121 are still there, but the fix needs to be rewritten for the new slider code. @umlaeute is it worth it for me to start it as a new PR targeted to this branch?

umlaeute commented:
just for the sake of my sanity: please comment on the relevant bugs reports (and possibly link to this PR from there) rather than the other way round.

e.g. https://github.com/pure-data/pure-data/pull/1633#issuecomment-1105869303 could go into https://github.com/pure-data/pure-data/issues/1638

ideally we would be able to create "bug-reports for PRs", but github doesn't offer this; so let's use the normal tickets.

but having micro-discussions for separate issues within this PR,just makes the conversation unreadable (for generally interested parties, potential mergers (@millerpuckette) and most importantly (for me :yum: ) yours sincerely)

porres commented:
sorry to polute this even more (we can delete comments later) but I tried creating a PR targeted to this branch, what did I do wrong? I loaded it on 'github' destop, made the change and it said _Authentication failed_ when I tried to push upstream, so I made these: https://github.com/pure-data/pure-data/pull/1639 & https://github.com/pure-data/pure-data/pull/1640 (see screenshot of my problem in https://github.com/pure-data/pure-data/pull/1640#issuecomment-1105865250). But I see @Ant1r could easily make a PR here. What are the steps?

Spacechild1 commented:
> What are the steps?

1. Create a new local branch and commit your changes
2. Push changes to your fork
3. Make a PR from your fork to this repo

I do not use the GitHub desktop app, so I can't really help you there. With the git command line app, it is pretty straight-forward...

> when I tried to push upstream

What is "upstream" in your case? Usually, "origin" would be your fork and "upstream" would be the original repo. In that case, you would rather push to "origin".

umlaeute commented:
Am 22. April 2022 13:48:57 MESZ schrieb porres ***@***.***>:
> But I see @Ant1r could easily make a PR here. What are the steps? 
>

The same as always: create your branch and push to your repository.
Then go the https://GitHub.com/pure-data/pure-data to create a new PR, but set the target branch to *this* one.

If all fails, we can retroactively change the target...

porres commented:
tried another strategy using the github site and now I believe this is ok and it will work --> https://github.com/pure-data/pure-data/pull/1642 please confirm

umlaeute commented:
no idea why github shows this as unmergeable: i've just merged the `master` branch into this branch and resolved all the conflicts manually with 8812c1b090b0d41ed6b9e7ec9d862acbb2213e81

nevertheless, this PR is also included in #1570 and i would suggest to merge #1570 instead.
